contrib/epee/src/wire/json/read.cpp

```diff
@@ -0,0 +1,452 @@
+#include "serialization/wire/error.h"
+#include "serialization/wire/json/error.h"
+
+namespace
+{
+  //! Maximum number of bytes to display "near" JSON error.
+  constexpr const std::size_t snippet_size = 30;
+
+  struct json_default_reject : rapidjson::BaseReaderHandler<rapidjson::UTF8<>, json_default_reject>
+  {
+    bool Default() const noexcept { return false; }
+  };
+
+  //! \throw std::system_error by converting `code` into a std::error_code
+  [[noreturn]] void throw_json_error(const epee::span<const std::uint8_t> source, const rapidjson::Reader& reader, const wire::error::schema expected)
+  {
+    const std::size_t start = std::min(source.size(), reader.GetErrorOffset());
+    const std::size_t end = start + std::min(snippet_size, source.size() - start);
+
+    const boost::string_ref text{reinterpret_cast<const char*>(source.data()) + start, end - start};
+    const rapidjson::ParseErrorCode parse_error = reader.GetParseErrorCode();
+    switch (parse_error)
+    {
+    default:
+      WIRE_DLOG_THROW(wire::error::rapidjson_e(parse_error), "near \"" << text << '"');
+    case rapidjson::kParseErrorNone:
+    case rapidjson::kParseErrorTermination: // the handler returned false
+      break;
+    }
+    WIRE_DLOG_THROW(expected, "near \"" << text << '"');
+  }
+}
+
+namespace wire
+{
+  struct json_reader::rapidjson_sax
+  {
+    union
+    {
+      bool boolean;
+      std::intmax_t integer;
+      std::uintmax_t unsigned_integer;
+      double number;
+    } value;
+
+    std::string* const string_;
+    const error::schema expected_;
+    error::schema actual_;
+    bool negative;
+
+    rapidjson_sax(error::schema expected) noexcept
+      : string_(nullptr), expected_(expected), actual_(error::schema::maximum_depth), negative(false)
+    {}
+
+    explicit rapidjson_sax(std::string& string, error::schema expected) noexcept
+      : string_(std::addressof(string)), expected_(expected), actual_(error::schema::maximum_depth), negative(false)
+    {}
+
+    bool Null() noexcept
+    {
+      actual_ = error::schema::none;
+      return expected_ == error::schema::none;
+    }
+
+    bool Bool(bool i) noexcept
+    {
+      value.boolean = i;
+      actual_ = error::schema::boolean;
+      return expected_ == error::schema::boolean || expected_ == error::schema::none;
+    }
+
+    bool Int(int i) noexcept
+    {
+      return Int64(i);
+    }
+    bool Uint(unsigned i) noexcept
+    {
+      return Uint64(i);
+    }
+    bool Int64(std::int64_t i) noexcept
+    {
+      negative = true;
+      switch(expected_)
+      {
+      default:
+        return false;
+      case error::schema::none:
+      case error::schema::integer:
+        value.integer = i;
+	actual_ = error::schema::integer;
+        break;
+      case error::schema::number:
+        value.number = i;
+	actual_ = error::schema::number;
+        break;
+      }
+      return true;
+    }
+    bool Uint64(std::uint64_t i) noexcept
+    {
+      switch(expected_)
```

Comment:
Is `negative` supposed to be set here?

Reply:
I guess it doesn't matter if you only use `rapidjson_sax` exactly once

Reply:
No, because the actual value is not negative.